### PR TITLE
blob: add support for As for returned errors

### LINKED ISSUE #750: blob: add 'As' for returned errors
Per @jba on #748 : that will provide an escape hatch for provider-aware error testing. See go.googlesource.com/proposal/+/master/design/go2draft-error-inspection.md.

### PR DESCRIPTION
**Pull request by vangent**
This PR adds `blob.ErrorAs`, which ends up calling `driver.Bucket.ErrorAs`. It works in a similar way to other `As` functions, with the slight difference that it is a top-level function (e.g., `blob.ErrorAs(err, i)` rather than `foo.As(i)`).

I think that if http://go.googlesource.com/proposal/+/master/design/go2draft-error-inspection.md were done, then we could change `blob` functions to all return a `blob.Error`, and add `blob.Error.As`; this would be nice. I don't think we can do it now, because at least `Reader` and `Writer` need to return `io.EOF` errors sometimes, not wrapped `blob.Error`.

Fixes #750.

### CODE REVIEW

File: blob/gcsblob/gcsblob.go [hunk 1]
```diff
@@ -269,6 +270,20 @@ func (b *bucket) As(i interface{}) bool {
 	return true
 }
 
+// As implements driver.ErrorAs.
+func (b *bucket) ErrorAs(err error, i interface{}) bool {
```
[the remark this answers is not in the document]
Comment (vangent) [reply blob/gcsblob/gcsblob.go]: Not sure I follow. The underlying type of err is `*googleapi.Error`. Here I've exposed it as `googleapi.Error`, so the user wants that type. Are you suggesting that we expose it as `*googleapi.Error` instead?

File: blob/s3blob/s3blob.go [hunk 2]
```diff
@@ -307,6 +308,20 @@ func (b *bucket) As(i interface{}) bool {
 	return true
 }
 
+// As implements driver.ErrorAs.
+func (b *bucket) ErrorAs(err error, i interface{}) bool {
+	switch v := err.(type) {
+	case awserr.Error:
+		p, ok := i.(*awserr.Error)
```
[the remark this answers is not in the document]
Comment (vangent) [reply blob/s3blob/s3blob.go]: Done.

File: blob/gcsblob/gcsblob.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (jba) [reply blob/gcsblob/gcsblob.go]: Yes, the type that implements the `error` interface is `*googleapi.Error`. If you want to get back one of those, you need to pass in a `**googleapi.Error`.

[the remark this answers is not in the document]
Comment (vangent) [reply blob/gcsblob/gcsblob.go]: OK, done.

**Pull request merged**